Unattended Gathering

matthewhillier opened the pull request:
Currently built on-top of the Character Skipping branch, this is a first-pass version of taking away the need for the Finish ID parameter and allow the gatherer to stop itself when it hits a sequential block of empty character ID's (currently set to 200).

Finish ID is still present for the purposes of testing, however it can now be left out completely and the gatherer will continue to process until it reaches the stop condition.

The current implementation is a little rough - it gathers in batches of 1000, stopping after the thread pool has died and then queries the DB for the stop condition. If the condition isn't met, then it recurses and carries on for the next 1000. The pause is fairly small, though can be affected if the web server is throttling requests causing the current batch to take a while before being assessed.

I think the next step here will be to thread the code that loads the character id's into the gathering queue, and the code that checks for the stop conditions, separately so that they're happening in parallel and we don't need to have an unnecessary pause to check on the state of the DB.

Fun times 😄

Currently tested at the top-end of the ID's in the website, identifying 21026701 as the last character ID at the last time of running. Funnily enough, new characters are being created quite frequently, so everytime I've run this the number increments slightly, but so far the stopping logic seems to be working 😄

Curiously, I did note some big gaps in deleted characters lower-down in the ID range. At first I thought it might indicate that SE aren't creating character ID's in order, but then later in the database I found what looks to be large chunks of bot characters (all Lallafel's, all level 50 thaumaturge or conjourer, all associated with free-companies with gibberish names). The batch sizes are similar to the gaps below, so I'm hypothesising SE does a bit of a clean-up exercise every now and then :)

matthewhillier committed:
Refactored executor loop into a simple recursive method that gathers characters, 1000 at a time.

matthewhillier committed:
Implemented a basic check into each gathering iteration that looks for the size of the gap between the highest known "valid" ID number, and the highest gathered ID number in the database, and stops gathering if it breaches a threshold of 200.
Integration test added to verify the stopping condition logic.

matthewhillier committed:
Implemented a first-pass try at stopping gathering activity after a block of 200 empty characters found at the top of the database.

matthewhillier committed:
Fixed unit test looking for the Finish option mandatory check.

matthewhillier commented:
Note, this is a rebased branch of https://github.com/XIVStats/XIVStats-Gatherer-Java/pull/29

Pricetx commented:
Hey,

Thanks for the work on this. I tried launching it and encountered an error. I don't have time tonight to look at the cause, but i'll have a look tomorrow. In the mean time, I thought you might be interested to see the output in the meantime: https://pastebin.com/KcrqYtYx

matthewhillier commented:
Hey @Pricetx, it looks like this is because there's command-line params being passed-in that were removed in the previous branch (in this case, the `-m` param for mounts).

See my comments [here](https://github.com/XIVStats/XIVStats-Gatherer-Java/pull/28#issuecomment-364646156) for the params that are no-longer available.

I'll raise a Bug to try and make this more obvious, as the error output mixing Spring Boot with command-line options doesn't look very obvious this way.

Hope that helps!

ReidWeb commented:
Any reason the mount list was removed? I think @pricetx may still use it for one of his tests.... Didn't know this myself until yesterday

matthewhillier commented:
Hey both, sorry - I guess I didn't make that change obvious enough!

I haven't removed the mount list, rather it will now **always** store the mounts (and minions, and timestamps, etc).

Those options were only used in the original DAO code that got replaced in the new branch, so in the interests of "_keeping it simple_" (or at least trying to) I've let the new data-access code always store the values and removed the options which made it "_don't store unless specified_".

I think the impact should only be on the DB storage size in test environments, as we're already reading-in the data from the lodestone and the live census data always includes these fields anyway. I haven't come across any performance impact as of yet.

Having said that, if you guys do need the ability to switch these fields off, then I can find another way to implement these options back in.

matthewhillier committed:
Reimagined the gathering behaviour to run completely in parallel threads:

- GathererTasks are now queued upto 1000 at a time
- LevemeteTask manages topping-up the GathererTask queue, limited by the user-defined finish ID (where specified)
- GatheringLimiterTask monitors for the large gaps in valid players and issues a halt on new GathererTasks when the limit is breached

TODO: Add unit tests for the new classes.

ReidWeb commented:
Should be sufficient to not have the option, we were always running with it off anyways. Fine with it unless @Pricetx has any objections.

Pricetx commented:
Hi Matt,

Thanks for the info on that issue, I have no problem with mounts always being gathered, it's just good to know that I can still rely on it being there!

I have tried reducing the arguments down to just -s and -f, and I now get this:
Caused by: org.springframework.dao.InvalidDataAccessApiUsageException: [Assertion failed] - this argument is required; it must not be null; nested exception is java.lang.IllegalArgumentException: [Assertion failed] - this argument is required; it must not be null

I figured it might be a problem with config.xml, so I renamed it to have it generate a new one, but running it doesn't actually generate a new config.xml. I grabbed a fresh copy of config.xml from your branch but it looked the same and still produced the same error.

I tried deleting the tblplayers table, which resulted in:
could not extract ResultSet
Table 'ffxivcensus_dev.tblplayers' doesn't exist

I then tried deleting DATABASECHANGELOG and DATABASECHANGELOGLOCK, which caused this:
Caused by: java.lang.IllegalArgumentException: [Assertion failed] - this argument is required; it must not be null

That final error looks like i'm back to square one in terms of errors. I can see that it did at least recreate the tables if I dropped all three of them.

Let me know if you require the full stack trace to debug further. I might also recommend trying to run the version that was built on Peter's Jenkins server, as it could be an issue with the jar file itself.

Thanks!

matthewhillier commented:
Hey @Pricetx - I found it! There was a potential Null-Pointer scenario which was only showing up immediately after a fresh upgrade or on creation of a new DB. The fix is in Pull Request #32.

Pricetx commented:
Hi Matthew,

Thanks for your patience. I've just grabbed  the latest build of this branch from jenkins and kicked it off in the test environment. It's running IDs 19,000,000 to 20,000,000. Looking at the output it looks like it's running alright. I've configured it to work with the latest version of the PHP in master, so hopefully we get positive results.

It should be finished within 24 hours, and if successful will be on https://dev.ffxivcensus.com/

As an aside, within the next couple of weeks I hope to have a new rancher environment up and running, so I should be able to create you an account and you can test and deploy directly.

Pricetx commented:
The testing has passed, though i'm not sure why the Jenkins build is now marked as failed. I will have to have a look.

Pricetx commented:
Found the problem:

testRunAdvancedOptions(com.ffxivcensus.gatherer.GathererControllerIT)  Time elapsed: 0.042 sec  <<< ERROR!
java.lang.Exception: org.springframework.dao.InvalidDataAccessResourceUsageException: could not extract ResultSet; SQL [n/a]; nested exception is org.hibernate.exception.SQLGrammarException: could not extract ResultSet

Going to have a quick look.

matthewhillier committed:
Code cleanup

Removed some unused code and made a tweak to improve testability.

matthewhillier committed:
Added a unit test for the Console class to test impact on CodeCov stats.

matthewhillier committed:
Additional coverage for safe execution of Help Docs

matthewhillier committed:
Further improvements to unit tests

matthewhillier committed:
Trying additons to the IntegrationTests that may be too sensitive with DB content. TBD.

matthewhillier committed:
Moved the *Task classes into a distinct task package

matthewhillier committed:
Turn off the offending test in gatherer controller for now.

matthewhillier committed:
Some code cleanup from sonarlint.

matthewhillier committed:
New Unit Test and some code smells cleanup

matthewhillier committed:
Added further unit test coverage

matthewhillier committed:
Added last test case for Task classes

matthewhillier committed:
Improved coverage across ConfigurationBuilder

matthewhillier committed:
Numbers bits of code cleanup

- Using SonarLint plugin on Eclipse to get the majority of these uncovered

matthewhillier commented:
Think I'm done with this branch now. I've been through code coverage stats with a fine tooth comb and added a lot more unit testing for the new classes. Looks like the last CodeCov report failed to upload, either their site was unavailable or I've made too many pushes in a short period of time. Feel free to re-run the jenkins build in a while to verify.

Hopefully the next time you run this, it should be more solid :)

matthewhillier committed:
Fixed a formatting issue in the readme file

matthewhillier commented:
Hmm.. something weird is going on here. The above failure is linking to build 4 in Jenkins, while the most recent build is [14](https://ci.reidweb.com/blue/organizations/jenkins/XIVStats%2FXIVStats-Gatherer-Java/detail/unattended_gathering/14/pipeline), which has passed... 😕

The failure email I got back from Jenkins contained the following:

> `[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project XIVStats-Gatherer-Java: Fatal error compiling: invalid target release: 1.8 -> [Help 1]`

This doesn't line-up with build 4 in the public builds, so it looks like "_something else_" is trying to build the project and isn't able. Whatever this is, it's reporting itself as `ci.reidweb.com`, but clearly isn't.

ReidWeb commented:
Apologies, I'll endeavour to take a look at the build issue some time this weekend, perhaps next week, hard to say - busy weeks ahead.

ReidWeb commented on src/test/resources/config/test_config.xml:
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<config>
+    <jdbc>
+        <url>mysql://testbox:3306</url>
```
Hmm...what's this?

ReidWeb commented:
CodeCov report can be ignored - i'll manually circumvent that.

ReidWeb commented on src/test/resources/config/test_config.xml:
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<config>
+    <jdbc>
+        <url>mysql://testbox:3306</url>
```
Hmm what's this?

matthewhillier replied:
Thanks, Peter.

This is a private test-resource copy of the config that's used exclusively by the ConifgurationBuilderTest unit test [here](https://github.com/XIVStats/XIVStats-Gatherer-Java/pull/30/files#diff-32f50191e9fba98cccc42dfdd2795d2dR41), to verify the behaviour of the XML configuration code. It doesn't get used in any actual DB configuration for the full-stack integration tests beyond this.

I've placed it under the `test/resources` folder so that it's only available during testing builds and never during final JAR generation, and it uses a separate name so that it won't get picked-up by any of the other config-loading calls.

ReidWeb reviewed (commented):
LGTM just want a bit of clarity on what that config is for - assuming just defaults that user's would fill in

ReidWeb commented:
Hey, looks like you need to rebase for I can merge?

matthewhillier commented:
Ok, after a few arguments with rebase and the github desktop client, we should now have a nicely rebased branch :)

matthewhillier commented:
🥂
